Milestone 5: Friendship v1

PhillipUg opened the pull request:
### In this feature, i have implemented the following:
- Enabled users to send and receive friend requests
- Enabled users to accept or reject friend requests from other users
- Enabled the 'Invite to frienship' button both at users path and user page

PhillipUg committed:
added Hirb gem

PhillipUg committed:
generated friendship model and added validations and associations

PhillipUg committed:
added invite action to users controller

PhillipUg committed:
Added 'add_friend' class method to Friendship model

PhillipUg committed:
Added current_user's name to navbar

PhillipUg committed:
Added invite-to-friendship buttons to both users' list and single user page

PhillipUg committed:
changed confirmed column name to state & its type from boolean to string

PhillipUg committed:
added 'accept' route

PhillipUg committed:
enable users to see friend requests and accept or reject buttons

PhillipUg committed:
added helper methods in user model

PhillipUg committed:
added 'accept' action in users controller

PhillipUg committed:
added reject route

PhillipUg committed:
added 'my friends', reject link, to user show page

PhillipUg committed:
removed current user from showing up among all users

PhillipUg committed:
added 'my profile' link in navbar

PhillipUg committed:
modified 'add friend' method for mutual frienship

PhillipUg committed:
modified invite action for mutual frienship and added reject action

PhillipUg committed:
modified confirm friend helper method

PhillipUg committed:
removed the unused helpers

PhillipUg committed:
enabled users to only see requets they recieve not the ones they send

PhillipUg committed:
added and configured rspec & capybara

PhillipUg committed:
Added shoulda-matchers, launchy, faker and factory-bot gems

PhillipUg committed:
Installed and configured rspec

PhillipUg committed:
added model specs

PhillipUg committed:
added request specs

PhillipUg committed:
added feature specs

PhillipUg committed:
added 'id' attributes to form fields in views to make them accessible to capybara

PhillipUg committed:
removed the 'dependent destroy' validation option since we wont be deleting users or posts

PhillipUg committed:
added user factories

PhillipUg committed:
added '=nil' option to silence warnings in terminal

PhillipUg committed:
added class attributes to like and dislike buttons for easy testing

PhillipUg committed:
corrected rubocop errors

PhillipUg committed:
rubocop errors

PhillipUg committed:
enabled users to see the requests they send to other users

PhillipUg committed:
added some style to the buttons

PhillipUg committed:
more styles

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello @PhillipUg 
Thank you for the work done!
It's almost OK. However, there are some missing requirements

1. It's asked [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) that posts timeline includes posts written by logged-in user and their friends. However, it's not what you did. Kindly do it

2. I created two users **Alice Bar** and **Bob Bar**
Alice sent a friendship request to Bob. Here is the result from the database
```rb
2.7.0 :001 > Friendship.all
  Friendship Load (0.6ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 1, user_id: 2, friend_id: 1, state: "pending", created_at: "2020-06-14 09:18:58", updated_at: "2020-06-14 09:18:58">, #<Friendship id: 2, user_id: 1, friend_id: 2, state: "request-sent", created_at: "2020-06-14 09:18:58", updated_at: "2020-06-14 09:18:58">]> 
2.7.0 :002 > 
```
Which is not OK at this level of the project. The `Milestone 5` is about one line create when Friendship request is sent.

Next, I logged in as Bob to answer to the friendship request sent by Alice. But nothing!! Here is what I got
![Screenshot_2020-06-14_10-23-33](https://user-images.githubusercontent.com/8848104/84589808-22dca880-ae29-11ea-977b-ce7ad9703d23.png)
You could notice that there is nothing concerning **Accepting** or **Rejecting** the Friendship. 
If I clicked on the **Invite to Friendship** the result will be bad.

3- NO LOGIC IF VIEW. Remove if/else statement from your views. Create the helpers views methods for that.

Kindly make the corresponding changes and submit a new code review request
Thanks

Mael FOSSO

PhillipUg committed:
added likes & comments count in views

PhillipUg committed:
modified my_friends instance methods to give accurate output

PhillipUg committed:
modified add_friend class method output

PhillipUg committed:
modified accept, reject and invite actions in user controller

PhillipUg committed:
modified timeline posts to show only posts by users and their friends

PhillipUg committed:
corrected rubocop errors

PhillipUg committed:
added database cleaner gem

PhillipUg committed:
corrected failing specs as a result of recent changes

PhillipUg committed:
rubocop errors

PhillipUg commented:
#### Made the following fixes:
- filtered timeline posts to show posts from only users and their friends
- corrected issue with database double records
- modified views to show accept and reject buttons on the users_path

PhillipUg committed:
added feature specs for accepting and rejecting requests

Torres-ssf reviewed (approved):
### Project Approved :+1: 
Hi there @PhillipUg 
Great job on this project :clap: 

**FOR THE NEXT MILESTONE**
One of the requirements of the previous review was left behind. You should remove all logic from the views. You read more about in this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers).

Good luck in the next PR.

Happy coding :computer: 
**Sergio Torres**
